systemtest/infra_kafka.go

```diff
@@ -21,233 +21,203 @@ import (
+		Namespace:   kafkaNamespace,
+		PortMapping: "0:9093",
 	}
 	pf, err := pfReq.New(context.Background(), stopCh)
-	if err != nil {
-		return err
-	}
+	require.NoError(t, err)
+	wg.Add(1)
 	go func() {
+		defer wg.Done()
 		if err := pf.ForwardPorts(); err != nil {
-			logger().Fatalf("port forward terminated unexpectedly: %v", err)
+			t.Fatalf("port forwarder terminated unexpectedly: %v", err)
 		}
 	}()
-
-	RegisterDestroy(ns+service+mapping, func() {
-		srv.Close()
+	t.Cleanup(func() {
 		close(stopCh)
+		wg.Wait()
 	})
 
 	// wait for port forward to be ready
-	select {
-	case <-ctx.Done():
-		return fmt.Errorf("failed while waiting for port forward to be ready: %w", err)
-	case <-pf.Ready:
-	}
-	return nil
-}
-
-func newKafkaTLSDialer() *tls.Dialer {
-	return &tls.Dialer{
-		NetDialer: &net.Dialer{Resolver: mockResolver},
-		Config:    &tls.Config{InsecureSkipVerify: true},
-	}
+	<-pf.Ready
+	ports, err := pf.GetPorts()
+	require.NoError(t, err)
+	fmt.Println(ports[0])
+	return fmt.Sprintf("127.0.0.1:%d", ports[0].Local)
 }
 
-// KafkaBrokers returns the Kafka brokers to use for tests.
-func KafkaBrokers() []string {
-	brokersMu.RLock()
-	defer brokersMu.RUnlock()
-	return append([]string{}, kafkaBrokers...)
+// NewKafkaManager returns a new kafka.Manager for the configured broker.
+//
+// The manager is configured to create topics with 1 partition and to
+// retain messages for 1 hour.
+func NewKafkaManager(t testing.TB) (*kafka.Manager, error) {
+	return kafka.NewManager(kafka.ManagerConfig{
+		CommonConfig:   KafkaCommonConfig(t),
+		PartitionCount: 1,
+		TopicConfigs: map[string]string{
+			"retention.ms": strconv.FormatInt(time.Hour.Milliseconds(), 10),
+		},
+	})
 }
 
-// SetKafkaBrokers sets the kafka brokers.
-func SetKafkaBrokers(brokers ...string) {
-	brokersMu.Lock()
-	defer brokersMu.Unlock()
-	kafkaBrokers = append([]string{}, brokers...)
+// KafkaCommonConfig returns a kafka.CommonConfig suitable for connecting to
+// the configured Kafka broker in tests.
+//
+// When Kafka is running in Kubernetes, this will take care of forwarding the
+// necessary port(s) to connect to the broker, and clean up on test completion.
+func KafkaCommonConfig(t testing.TB) kafka.CommonConfig {
+	commonConfig := kafka.CommonConfig{
+		Brokers: append([]string{}, kafkaBrokers...),
+		Logger:  logger().Desugar().Named("kafka"),
+	}
+	if len(commonConfig.Brokers) == 0 {
+		brokerAddress := portforwardKafka(t)
+		tlsDialer := &tls.Dialer{
+			// running locally, should be fast
+			NetDialer: &net.Dialer{Timeout: time.Second},
+			Config:    &tls.Config{InsecureSkipVerify: true},
+		}
+		commonConfig.Brokers = []string{brokerAddress}
+		commonConfig.Dialer = func(ctx context.Context, network, addr string) (net.Conn, error) {
+			// The advertised broker address is only reachable within
+			// the Kubernetes cluster; replace it with the localhost
+			// port-forwarded address.
+			addr = brokerAddress
```

Comment:
ah nice, so this removes the need for the mockdns?

Reply:
yes, there's only one broker, so we can just ignore the advertised address